client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1035 @@
+                        extension_length -= kaa_aligned_size_get(notification_size);
+                        if (!notifications_count) {
+                            kaa_list_t *state_found = kaa_list_find_next(self->status->topic_states, &kaa_find_topic_state_by_id, &topic_id);
+                            if (!state_found) {
+                                kaa_list_t *new_topic_state = NULL;
+                                kaa_topic_state_t *state = (kaa_topic_state_t *) KAA_MALLOC(sizeof(kaa_topic_state_t));
+                                if (!state) {
+                                    KAA_FREE(notification);
+                                    return KAA_ERR_NOMEM;
+                                }
+                                state->sqn_number = seq_number;
+                                state->topic_id = topic_id;
+                                if (!self->status->topic_states) {
+                                    new_topic_state = kaa_list_create(state);
+                                } else {
+                                    new_topic_state = kaa_list_push_front(self->status->topic_states, state);
+                                }
+                                if (!new_topic_state) {
+                                    KAA_FREE(state);
+                                    return KAA_ERR_NOMEM;
+                                }
+                                self->status->topic_states = new_topic_state;
+                            } else {
+                                kaa_topic_state_t *state = (kaa_topic_state_t *)kaa_list_get_data(state_found);
+                                state->sqn_number = seq_number;
+                            }
+                        }
+                        err = kaa_notification_received(self, notification, &topic_id);
+                        notification->destroy(notification);
+                        if (err) {
+                            KAA_LOG_WARN(self->logger, err, "Failed to notify listeners.");
+                        }
+                    }
+                }
+                break;
+            }
+            case TOPICS: {
+                KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Received topics");
+                uint16_t topic_count = KAA_NTOHS(*((uint16_t *) reader->current));
+                KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Topic count is %u", topic_count);
+                reader->current += sizeof(uint16_t);
+                extension_length -= sizeof(uint16_t);
+                kaa_list_t *topics = NULL;
+                kaa_topic_t *topic = NULL;
+                kaa_list_t *new_topic = NULL;
+                while (topic_count--) {
+                   topic = (kaa_topic_t *) KAA_MALLOC(sizeof(kaa_topic_t));
+                   KAA_RETURN_IF_NIL(topic, KAA_ERR_NOMEM);
+                   topic->id = KAA_NTOHLL(*((uint64_t *) reader->current));
+                   reader->current += sizeof(uint64_t);
```

Comment:
You can reduce code amount by replacing
reader->current += sizeof(uint16_t);
extension_length -= sizeof(uint16_t);
by some appropriate function or macros.